pdns/fstrm_logger.cc

```diff
@@ -0,0 +1,110 @@
+#include <unistd.h>
+#include "fstrm_logger.hh"
+#include "config.h"
+#ifdef PDNS_CONFIG_ARGS
+#include "logger.hh"
+#define WE_ARE_RECURSOR
+#else
+#include "dolog.hh"
+#endif
+#include <sys/un.h>
+
+#define DNSTAP_CONTENT_TYPE		"protobuf:dnstap.Dnstap"
+
+FrameStreamLogger::FrameStreamLogger(const std::string socket_path): socket_path(socket_path)
+{
+  struct sockaddr_un local;
+  local.sun_family = AF_UNIX;
+  strcpy(local.sun_path, socket_path.c_str());
+
+  fwopt = fstrm_writer_options_init();
+  fstrm_res res = fstrm_writer_options_add_content_type(fwopt, DNSTAP_CONTENT_TYPE, sizeof(DNSTAP_CONTENT_TYPE) - 1);
+  if (res != fstrm_res_success) {
+#ifdef WE_ARE_RECURSOR
+    L<<Logger::Warning<<"Error: fstrm_writer_options_add_content_type failed: "<<res<<std::endl;
+#else
+    warnlog("Error: fstrm_writer_options_add_content_type failed: %s\n", res);
+#endif
+  }
+
+  uwopt = fstrm_unix_writer_options_init();
+  fstrm_unix_writer_options_set_socket_path(uwopt, local.sun_path);
+
+  writer = fstrm_unix_writer_init(uwopt, fwopt);
+  if (!writer) {
+#ifdef WE_ARE_RECURSOR
+    L<<Logger::Warning<<"Error: fstrm_unix_writer_init() failed."<<std::endl;
+#else
+    warnlog("Error: fstrm_unix_writer_init(%s) failed.\n", socket_path);
+#endif
+  }
+
+  iothr = fstrm_iothr_init(NULL, &writer);
+  if (!iothr) {
+#ifdef WE_ARE_RECURSOR
+    L<<Logger::Warning<<"Error: fstrm_iothr_init() failed."<<std::endl;
+#else
+    warnlog("Error: fstrm_iothr_init() failed.\n");
+#endif
+  }
+
+  ioqueue = fstrm_iothr_get_input_queue(iothr);
```

Comment:
Since `queueData()` can be called from several threads, the queue model should be set to `FSTRM_IOTHR_QUEUE_MODEL_MPSC` with `fstrm_iothr_options_set_queue_model()`